zaza/openstack/charm_tests/ceph/tests.py

```diff
@@ -1032,15 +1032,22 @@ def get_mon_count(self, unit):
         match = rx.search(result['Stdout'])
         return int(match.group(1))
 
+    def get_mon_units(self):
+        """Get a list of the ceph-mon units."""
+        return list(x.entity_id for x in zaza_model.get_units('ceph-mon'))
+
     def test_departure(self):
         """Test descaling of a ceph-mon unit."""
-        mons = list(x.entity_id for x in zaza_model.get_units('ceph-mon'))
+        mons = self.get_mon_units()
         mon_count = len(mons)
         logging.info('Monitor count is {}'.format(mon_count))
-        zaza_model.add_unit('ceph-mon')
-        zaza_model.block_until_unit_count('ceph-mon', mon_count + 1)
-        zaza_model.block_until_all_units_idle(model_name=self.model_name)
-        self.assertEqual(mon_count + 1, self.get_mon_count(mons[0]))
-        zaza_model.destroy_unit('ceph-mon', mons[0])
-        zaza_model.block_until_unit_count('ceph-mon', mon_count)
-        self.assertEqual(mon_count, self.get_mon_count(mons[1]))
+        zaza_model.add_unit('ceph-mon', wait_appear=True)
+        try:
+            logging.info('New monitor has been added')
+            new_mon = next(iter(set(self.get_mon_units()) - set(mons)))
+            zaza_model.wait_for_unit_idle(new_mon)
+            self.assertEqual(mon_count + 1, self.get_mon_count(mons[0]))
+            zaza_model.destroy_unit('ceph-mon', mons[0], wait_disappear=True)
+            self.assertEqual(mon_count, self.get_mon_count(mons[1]))
+        except Exception:
+            zaza_model.destroy_unit('ceph-mon', mons[0])
```

Comment:
It seems a bit odd to wrap this in an Exception that catches everything?  I don't understandas this is supposed to be a test?  Shouldn't it fail when something goes wrong?

Reply:
Definitely. The idea of that clause is that the unit that was added is still removed regardless. This is done so that any subsequent tests run with the expected number of ceph-mon units. It may seem a bit wasteful since this is in case of failure anyway, but I think it's useful so that other tests run in a non-polluted environment.

Reply:
I don't think I explained myself properly.  The test can't fail as any exceptions are caught, unless destroy_unit() fails.  I think you have missed of a "raise" statement after the destroy_unit() call to re-raise the exception and fail the test.

Reply:
ahhh, yes, you are right. I'll add the code to reraise the exception.